lib/staggered_kd_apply_xinv.cu

```diff
@@ -0,0 +1,254 @@
+  template <typename vFloatSpinor, typename vFloatGauge, int fineColor, int fineSpin>
+  void applyStaggeredKDBlock(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &Xinv, bool dagger)
+  {
+    //constexpr int coarseSpin = 2;
+    const int coarseDof = Xinv.Ncolor(); // / coarseSpin;
+
+    if (coarseDof == 48) { // dof w/in a KD block
+      applyStaggeredKDBlock<vFloatSpinor, vFloatGauge, fineColor, fineSpin, 48>(out, in, Xinv, dagger);
+    } else {
+      errorQuda("Unsupported number of Kahler-Dirac dof %d\n", Xinv.Ncolor());
+    }
+  }
+
+  // template on fine colors, spin
+  template <typename vFloatSpinor, typename vFloatGauge>
+  void applyStaggeredKDBlock(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &Xinv, bool dagger)
+  {
+    if (out.Ncolor() != in.Ncolor()) 
+      errorQuda("Ncolor %d and %d do not match", out.Ncolor(), in.Ncolor());
+
+    if (out.Nspin() != in.Nspin())
+      errorQuda("Nspin %d and %d do not match", out.Nspin(), in.Nspin());
+
+    if (out.Ncolor() == 3 && out.Nspin() == 1) {
+      applyStaggeredKDBlock<vFloatSpinor, vFloatGauge, 3, 1>(out, in, Xinv, dagger);
+    } else {
+      errorQuda("Unsupported (color, spin) = (%d, %d)", out.Ncolor(), out.Nspin());
+    }
+  }
+
+  // template on Xinv precision (only half and single for now)
+  template <typename vFloatSpinor>
+  void applyStaggeredKDBlock(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &Xinv, bool dagger)
+  {
+
+#if QUDA_PRECISION & 4
+    if (Xinv.Precision() == QUDA_SINGLE_PRECISION) {
+      applyStaggeredKDBlock<vFloatSpinor, float>(out, in, Xinv, dagger);
+    } else
+#endif
+#if QUDA_PRECISION & 2
+    if (Xinv.Precision() == QUDA_HALF_PRECISION) {
+      applyStaggeredKDBlock<vFloatSpinor, short>(out, in, Xinv, dagger);
+    } else
+#endif
+    {
+      errorQuda("Unsupported precision %d", Xinv.Precision());
+    }
+  }
+
+  // Applies the staggered KD block inverse to a staggered ColorSpinor
+  void ApplyStaggeredKahlerDiracInverse(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &Xinv, bool dagger)
+  {
+#if defined(GPU_STAGGERED_DIRAC)
+    // FIXME: This will eventually get replaced by a custom instantiate
+    // I just don't want to think about it for now.
+    auto location = checkLocation(out, in, Xinv);
+
+    if (location == QUDA_CPU_FIELD_LOCATION)
+      errorQuda("There is no support for applying the KD operator to CPU fields (yet)");
+
+    // the staggered KD block inverse can only be applied to a full field
+    if (out.SiteSubset() != QUDA_FULL_SITE_SUBSET || out.SiteSubset() != QUDA_FULL_SITE_SUBSET)
+      errorQuda("There is no meaning to applying the KD inverse to a single parity field");
+
+    // Instantiate based on ColorSpinor precision
+    // We don't have a constraint on the precision of Xinv matching
+    // the precision of the spinors.
+    auto precision = checkPrecision(out, in);
+
+#if QUDA_PRECISION & 8
```

Comment:
If you turn `applyStaggeredKDBlock` into a class, you can use `instantiatePrecision` here to clean up this `if` / `else` precision boilerplate

Reply:
We should discuss this on a call---this file has the quirk that Xinv can be stored in a separate precision from the ColorSpinorField, but that may just be a separate flavor of `instantiatePrecision`.

Reply:
That shouldn't matter, you just need to pass a field with the correct precision to be instantiated as the first argument to `instantiatePrecision`.  Subsequent field arguments can have different precisions which can be unwrapper later.

Reply:
Done